pkg/kubelet/images/pullmanager/mem_pullrecords.go

```diff
@@ -0,0 +1,280 @@
+	cache   *lru.Cache
+	maxSize int
+
+	// authoritative indicates if we can consider the cached records an
+	// authoritative source.
+	// False if the cache is evicted any records because it reached capacity, or
+	// if there were errors during its initialization.
+	authoritative atomic.Bool
+
+	// deletingKeys is used by the eviction function to distinguish between keys
+	// being explicitly deleted and keys being removed because the cache was too
+	// large or being cleared.
+	//
+	// This is only modified by lruCache.Delete().
+	deletingKeys sync.Map
+}
+
+func newLRUCache[K comparable, V any](size int) *lruCache[K, V] {
+	c := lru.New(size)
+	l := &lruCache[K, V]{
+		maxSize:      size,
+		cache:        c,
+		deletingKeys: sync.Map{},
+	}
+	if err := c.SetEvictionFunc(func(key lru.Key, _ any) {
+		if _, shouldIgnore := l.deletingKeys.Load(key); shouldIgnore {
+			return
+		}
+		// any eviction makes our cache non-authoritative
+		l.authoritative.Store(false)
+	}); err != nil {
+		panic(fmt.Sprintf("failed to set eviction function to the LRU cache: %v", err))
+	}
+	return l
+}
+
+func (c *lruCache[K, V]) Get(key K) (*V, bool) {
+	value, found := c.cache.Get(key)
+	if !found {
+		return nil, false
+	}
+	if value == nil {
+		return nil, true
+	}
+	return value.(*V), true
+}
+
+func (c *lruCache[K, V]) Set(key K, value *V) { c.cache.Add(key, value) }
+func (c *lruCache[K, V]) Len() int            { return c.cache.Len() }
+func (c *lruCache[K, V]) Clear()              { c.cache.Clear() }
+
+// Delete will prevent authoritative cache status changes.
+//
+// Must be called locked with an external write lock on `key`.
+func (c *lruCache[K, V]) Delete(key K) {
+	c.deletingKeys.Store(key, struct{}{})
+	defer c.deletingKeys.Delete(key)
+	c.cache.Remove(key)
+}
+
+// cachedPullRecordsAccessor implements a write-through cache layer on top
+// of another PullRecordsAccessor
+type cachedPullRecordsAccessor struct {
+	delegate PullRecordsAccessor
+
+	intentsLocks       *StripedLockSet
+	intents            *lruCache[string, kubeletconfiginternal.ImagePullIntent]
+	pulledRecordsLocks *StripedLockSet
+	pulledRecords      *lruCache[string, kubeletconfiginternal.ImagePulledRecord]
+}
+
+func NewCachedPullRecordsAccessor(delegate PullRecordsAccessor, intentsCacheSize, pulledRecordsCacheSize, stripedLocksSize int32) *cachedPullRecordsAccessor {
+	intentsCacheSize = min(intentsCacheSize, 1024)
+	pulledRecordsCacheSize = min(pulledRecordsCacheSize, 2000)
+
+	c := &cachedPullRecordsAccessor{
+		delegate: delegate,
+
+		intentsLocks:       NewStripedLockSet(stripedLocksSize),
+		intents:            newLRUCache[string, kubeletconfiginternal.ImagePullIntent](int(intentsCacheSize)),
+		pulledRecordsLocks: NewStripedLockSet(stripedLocksSize),
+		pulledRecords:      newLRUCache[string, kubeletconfiginternal.ImagePulledRecord](int(pulledRecordsCacheSize)),
+	}
+	// warm our caches and set authoritative
+	_, _ = c.ListImagePullIntents()
```

Comment:
should errors be at least logged? Ignoring them completely seems to be creating a problem for debugging

Reply:
Sounds good, I added the error logging to https://github.com/kubernetes/kubernetes/pull/131882